File: src/com/leo/leetcode/algorithm/Q45.java
package com.leo.leetcode.algorithm;

public class Q45 {
    public static void main(String[] args) {
        System.out.println(new Q45().jump(new int[]{1, 1, 1, 1})); // 3
        System.out.println(new Q45().jump(new int[]{7, 0, 9, 6, 9, 6, 1, 7, 9, 0, 1, 2, 9, 0, 3})); // 2
        System.out.println(new Q45().jump(new int[]{2, 3, 1, 1, 4})); // 2
        System.out.println(new Q45().jump(new int[]{1, 2, 3})); // 2
        System.out.println(new Q45().jump(new int[]{2, 1})); // 1
        System.out.println(new Q45().jump(new int[]{1, 3})); // 1
        System.out.println(new Q45().jump(new int[]{1})); // 0
    }

    public int jump(int[] nums) {
        int out = 0, limit = 0, flag = 0;
        for (int i = 0; i < nums.length - 1; i++) {
            limit = Math.max(i + nums[i], limit);
            if (limit >= nums.length - 1) return out + 1;
            if (i == flag) {
                flag = limit;
                out++;
            }
        }
        return out;
    }
}

File: py/q0000/Q45.py
"""
 * 给定一个长度为 n 的 0 索引整数数组 nums。初始位置为 nums[0]。
 * 每个元素 nums[i] 表示从索引 i 向前跳转的最大长度。换句话说，如果你在 nums[i] 处，你可以跳转到任意 nums[i + j] 处:
 * 1、0 <= j <= nums[i] 
 * 2、i + j < n
 * 返回到达 nums[n - 1] 的最小跳跃次数。生成的测试用例可以到达 nums[n - 1]。
 * 提示:
 * 1、1 <= nums.length <= 10^4
 * 2、0 <= nums[i] <= 1000
 * 链接：https://leetcode.cn/problems/jump-game-ii/
"""
from typing import List


class Solution:

    def jump(self, nums: List[int]) -> int:
        l, r, n = 0, nums[0], len(nums)
        if n == 1: return 0
        ans = 1
        while r < n - 1:
            ans += 1
            l, r = r, max(i + nums[i] for i in range(l, r + 1))
        return ans


if __name__ == '__main__':
    # 0
    print(Solution().jump([0]))
    # 1
    print(Solution().jump([1, 3]))
    # 2
    print(Solution().jump([2, 3, 1, 1, 4]))
    # 2
    print(Solution().jump([2, 3, 0, 1, 4]))
